kafka/adminapi_test.go

```diff
@@ -0,0 +1,245 @@
+/**
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+)
+
+func testAdminAPIs(what string, a *AdminClient, t *testing.T) {
+	t.Logf("AdminClient API testing on %s: %s", a, what)
+
+	expDuration, err := time.ParseDuration("1.5s")
+	if err != nil {
+		t.Fatalf("%s", err)
+	}
+
+	confStrings := map[string]string{
+		"some.topic.config":  "unchecked",
+		"these.are.verified": "on the broker",
+		"and.this.is":        "just",
+		"a":                  "unit test"}
+
+	// Correct input, fail with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), expDuration)
```

Comment:
Just curious: Why does this time out? Cause the admin client's broker isn't set or defaults to something like localhost:9092 where no broker is running during this test and there's no response? You'd think the client would fail faster in that case if so.

Reply:
While the Go API does not expose it, librdkafka allows manually adding brokers after client instance inception (such as learnt from ZK or other external system), so even if no brokers are available at the time the admin API is called the admin state machine will wait for up to AdminOptions.RequestTimeout to send the request, handling broker connectivity issues, controller changes, and what not.

So yes, in this test we rely on the context to timeout before the actual underlying admin operation in librdkafka times out.